feat(drivable area expansion): do not over expand beyond the desired width

maxime-clem opened the pull request:
## Description

<!-- Write a brief description of this PR. -->
This PR improves the calculations of the drivable area expansion. Previously, the left and right bounds were expanded independently. With this PR, they are expanded together such that the resulting drivable area width is more accurately equals to the desired width.
Required launch PR: https://github.com/autowarefoundation/autoware_launch/pull/755

## Tests performed

<!-- Describe how you have tested this PR. -->
<!-- Although the default value is set to "Not Applicable.", please update this section if the type is either [feat, fix, perf], or if requested by the reviewers. -->
Psim
Evaluator (TIER IV INTERNAL LINK): 
- https://evaluation.tier4.jp/evaluation/reports/2491b8b3-717e-5cf8-be74-6d95c22181c1?project_id=x2_dev
- https://evaluation.tier4.jp/evaluation/reports/c4a4e2aa-55fd-59e4-b4b5-472c14457c5c?project_id=prd_jt

## Effects on system behavior

<!-- Describe how this PR affects the system behavior. -->
drivable area expansion more accurately achieve the desired minimum width.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [X] I've confirmed the [contribution guidelines].
- [X] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

maxime-clem commented on planning/behavior_path_planner/src/utils/drivable_area_expansion/drivable_area_expansion.cpp:
```diff
@@ -121,52 +121,77 @@ double calculate_minimum_lane_width(
+      arc_length +=
+        tier4_autoware_utils::calcDistance2d(bound[up_bound_idx - 1], bound[up_bound_idx]);
+      if (arc_length > extra_arc_length) break;
+      bound_expansions[up_bound_idx] = bound_expansions[bound_idx];
```
should do a max ?

maxime-clem replied:
Done: 59685d8c8799c023fcd463661e9a04e4fe31de7e

HansOersted commented on planning/behavior_path_planner/include/behavior_path_planner/utils/drivable_area_expansion/path_projection.hpp:
```diff
@@ -37,18 +37,15 @@ inline PointDistance point_to_segment_projection(
 
-  const auto cross = p2_vec.x() * p_vec.y() - p2_vec.y() * p_vec.x();
-  const auto dist_sign = cross < 0.0 ? -1.0 : 1.0;
-
```
Consider changing the comment on Line 29.

maxime-clem commented on planning/behavior_path_planner_common/src/utils/drivable_area_expansion/drivable_area_expansion.cpp:
```diff
@@ -121,52 +121,79 @@ double calculate_minimum_lane_width(
+  }
+}
+
+void apply_extra_arc_length(
+  Expansion & expansion, const std::vector<Point> & bound, const double extra_arc_length,
```
`extra_arc_length`  is confusing. Need a better name.

maxime-clem replied:
[d0490c4](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/d0490c461f9926cdbf3be61e252e28af68bf9b5e)

maxime-clem commented on planning/behavior_path_planner_common/src/utils/drivable_area_expansion/drivable_area_expansion.cpp:
```diff
@@ -121,52 +121,79 @@ double calculate_minimum_lane_width(
+  const auto original_expansions = bound_expansions;
+  for (auto path_idx = 0UL; path_idx < bound_indexes.size(); ++path_idx) {
+    const auto bound_idx = bound_indexes[path_idx];
+    auto arc_length = boost::geometry::distance(
+      bound_projections[path_idx].point, convert_point(bound[bound_idx + 1]));
+    for (auto up_bound_idx = bound_idx + 2; up_bound_idx < bound.size(); ++up_bound_idx) {
+      arc_length +=
+        tier4_autoware_utils::calcDistance2d(bound[up_bound_idx - 1], bound[up_bound_idx]);
+      if (arc_length > extra_arc_length) break;
+      bound_expansions[up_bound_idx] =
+        std::max(bound_expansions[up_bound_idx], original_expansions[bound_idx]);
+    }
```
See if this can be put in a lambda to be reused for the down direction

maxime-clem replied:
[76d713d](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/76d713d7e6603acfb6584b87e218171386ac6ed7)

maxime-clem commented on planning/behavior_path_planner_common/src/utils/drivable_area_expansion/drivable_area_expansion.cpp:
```diff
@@ -271,6 +296,48 @@ std::vector<double> calculate_smoothed_curvatures(
+    const auto original_width = expansion.left_projections[path_idx].distance +
+                                expansion.right_projections[path_idx].distance;
+    const auto expansion_dist = expansion.min_lane_widths[path_idx] - original_width;
+    if (expansion_dist > 0.0) {
```
replace with `continue` if negative

maxime-clem replied:
[76d713d](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/76d713d7e6603acfb6584b87e218171386ac6ed7)

danielsanchezaran commented on planning/behavior_path_planner_common/test/test_drivable_area_expansion.cpp:
```diff
@@ -58,7 +58,7 @@ TEST(DrivableAreaExpansionProjection, PointToSegment)
     Segment2d segment(Point2d(0.0, 0.0), Point2d(10.0, 0.0));
     const auto projection = point_to_segment_projection(query, segment.first, segment.second);
-    EXPECT_NEAR(projection.distance, -5.0, eps);
+    EXPECT_NEAR(projection.distance, 5.0, eps);
```
If the signs are added again, dont forget to test with negative values again.

danielsanchezaran reviewed (commented):
Thank you very much for this PR!
In general LGTM. The logic might be a little hard to follow when using the mapping indexes, so if you can put one or two comments in key places to explain the code it would be good, but not necessary. I leave it to your discretion.

maxime-clem commented on planning/behavior_path_planner_common/include/behavior_path_planner_common/utils/drivable_area_expansion/drivable_area_expansion.hpp:
```diff
@@ -57,6 +57,36 @@ void update_path_poses_and_previous_curvatures(
+  Expansion & expansion, const std::vector<Point> & bound, const double extra_arc_length,
+  const Side side);
+
+/// @brief calculate mappings between path poses and the given drivable area bound
```
[84fee4e](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/84fee4edc4ed9ef8879eae2945b48f5a60c0a282)

maxime-clem commented on planning/behavior_path_planner_common/include/behavior_path_planner_common/utils/drivable_area_expansion/path_projection.hpp:
```diff
@@ -26,7 +26,6 @@
 namespace drivable_area_expansion
 {
 /// @brief project a point to a segment
-/// @details the distance is signed based on the side of the point: left=positive, right=negative
```
The bug fix did not require using this function, so we can keep it unsigned.
[1eda3a0](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/1eda3a02170a697693d4e3500f3c57a483879542)

maxime-clem commented on planning/behavior_path_planner_common/src/utils/drivable_area_expansion/drivable_area_expansion.cpp:
```diff
@@ -271,6 +296,48 @@ std::vector<double> calculate_smoothed_curvatures(
+        std::min(max_left_expansions[left_bound_idx], max_left_expansions[left_bound_idx + 1]);
+      const auto right_limit =
+        std::min(max_right_expansions[right_bound_idx], max_right_expansions[right_bound_idx + 1]);
+      if (expansion_dist / 2.0 >= left_limit) {
```
Done: [76d713d](https://github.com/autowarefoundation/autoware.universe/pull/5640/commits/76d713d7e6603acfb6584b87e218171386ac6ed7)

maxime-clem commented:
@danielsanchezaran thank you for your review. I addressed all comments.
I had to rename one parameter so there is now a PR for autoware_launch: https://github.com/autowarefoundation/autoware_launch/pull/755
Please approve it as well. Thank you.
